smtlib-backends-process/tests/EdgeCases.hs

```diff
@@ -0,0 +1,47 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module EdgeCases (edgeCases) where
+
+import Control.Monad (when)
+import Data.ByteString.Builder (Builder)
+import SMTLIB.Backends as SMT
+import qualified SMTLIB.Backends.Process as Process
+import System.IO (hGetContents', hReady)
+import System.Process.Typed (getStdout)
+import Test.Tasty
+import Test.Tasty.HUnit
+
+edgeCases :: [TestTree]
+edgeCases =
+  [ testCase "Piling up stopping procedures" pileUpStops,
+    testCase "Sending an empty command" emptyCommand,
+    testCase "Sending a command expecting no response" commandNoResponse
+  ]
+
+-- | It's possible to accumulate procedures that stop the backend without
+-- hanging or crashing the program.
+pileUpStops :: IO ()
+pileUpStops = Process.with Process.defaultConfig $ \handle -> do
+  let backend = Process.toBackend handle
+  SMT.send_ backend "(exit)"
+  _ <- Process.close handle
+  Process.kill handle
+
+-- | Upon processing an empty command, the backend will not respond.
+emptyCommand :: IO ()
+emptyCommand = checkNoResponse ""
+
+-- | Upon processing a command producing no output, the backend will not
+-- respond, not even with an empty line.
+commandNoResponse :: IO ()
+commandNoResponse = checkNoResponse "(set-option :print-success false)"
+
+checkNoResponse :: Builder -> IO ()
+checkNoResponse cmd = Process.with Process.defaultConfig $ \handle -> do
+  let backend = Process.toBackend handle
+      stdout = getStdout $ Process.process handle
+  SMT.send_ backend cmd
+  responseAvailable <- hReady stdout
```

Comment:
This test is doomed to fail intermittently if the backend does respond. You would have to insert a delay to decrease the chance of the test passing when the response is coming late. There is no good delay to choose, as it depends on where the test is run, and tests would be artificially taking longer to run without doing any useful work.

Perhaps it would be fine to just document the behavior here without actually trying to execute it.

Reply:
Another option is to send another command that produces an output such as `(check-sat)` and check that there's no extra new line at the beginning of the response.

Reply:
https://github.com/tweag/smtlib-backends/pull/53/commits/0014c19ca589f9c31985daa0cb27292c7726ddbb